NAME - RIZA JAMES PETER

Group A : Data Science

ASSIGNMENT 1. Data Wrangling, I

Perform the following operations using Python on any open source dataset (e.g., data.csv)
1. Import all the required Python Libraries.
2. Locate an open source data from the web (e.g., https://www.kaggle.com). Provide a clear
 description of the data and its source (i.e., URL of the web site).
3. Load the Dataset into pandas dataframe.
4. Data Preprocessing: check for missing values in the data using pandas isnull(), describe()
function to get some initial statistics. Provide variable descriptions. Types of variables etc.
Check the dimensions of the data frame.
5. Data Formatting and Data Normalization: Summarize the types of variables by checking
the data types (i.e., character, numeric, integer, factor, and logical) of the variables in the
data set. If variables are not in the correct data type, apply proper type conversions.
6. Turn categorical variables into quantitative variables in Python.
In addition to the codes and outputs, explain every operation that you do in the above steps and
explain everything that you do to import/read/scrape the data set.


In [ ]:
#Step1 :IMPORTING REQUIRED LIBRARIES
import pandas as pd
import seaborn as sns

Step 2:Locate an open source data from the web (e.g., https://www.kaggle.com). Provide a clear description of the data and its source (i.e., URL of the web site).
Here I have used the dataset from seaborn library itself to load the dataset

In [ ]:
#Step 3: Loading the dataset into pandas dataframe
df =sns.load_dataset('titanic')

In [ ]:
df.head()

survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True

#step 4 :Data Preprocessing:
check for missing values in the data using pandas isnull(), describe() function to get some initial statistics. Provide variable descriptions. Types of variables etc. Check the dimensions of the data frame.

In [ ]:
#step 4 a:Checking for missing values
df.isnull().sum()

survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64

This returns the count of number of null present in each column.
Age has 177, deck has 688 and embark_town has 2 null values.

In [ ]:
#Step 4 b:Describing the dataset Get initial statistics
df.describe()

survived      pclass         age       sibsp       parch        fare
count  891.000000  891.000000  714.000000  891.000000  891.000000  891.000000
mean     0.383838    2.308642   29.699118    0.523008    0.381594   32.204208
std      0.486592    0.836071   14.526497    1.102743    0.806057   49.693429
min      0.000000    1.000000    0.420000    0.000000    0.000000    0.000000
25%      0.000000    2.000000   20.125000    0.000000    0.000000    7.910400
50%      0.000000    3.000000   28.000000    0.000000    0.000000   14.454200
75%      1.000000    3.000000   38.000000    1.000000    0.000000   31.000000
max      1.000000    3.000000   80.000000    8.000000    6.000000  512.329200

In [ ]:
#Step 4c: Variable descriptions and types
df.dtypes

survived          int64
pclass            int64
sex              object
age             float64
sibsp             int64
parch             int64
fare            float64
embarked         object
class          category
who              object
adult_male         bool
deck           category
embark_town      object
alive            object
alone              bool
dtype: object

In [ ]:
#Step 4 d: Provide variable descriptions
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 15 columns):
 #   Column       Non-Null Count  Dtype   
---  ------       --------------  -----   
 0   survived     891 non-null    int64   
 1   pclass       891 non-null    int64   
 2   sex          891 non-null    object  
 3   age          714 non-null    float64 
 4   sibsp        891 non-null    int64   
 5   parch        891 non-null    int64   
 6   fare         891 non-null    float64 
 7   embarked     889 non-null    object  
 8   class        891 non-null    category
 9   who          891 non-null    object  
 10  adult_male   891 non-null    bool    
 11  deck         203 non-null    category
 12  embark_town  889 non-null    object  
 13  alive        891 non-null    object  
 14  alone        891 non-null    bool    
dtypes: bool(2), category(2), float64(2), int64(4), object(5)
memory usage: 80.7+ KB


In [ ]:
df = df.dropna(subset=['embark_town'])

In [ ]:
deck_values = df['deck'].unique()
print("Unique values in the 'deck' column:")
print(deck_values)

Unique values in the 'deck' column:
[NaN, 'C', 'E', 'G', 'D', 'A', 'B', 'F']
Categories (7, object): ['A', 'B', 'C', 'D', 'E', 'F', 'G']


In [ ]:
df.shape

(889, 15)

This indicates that there are 889 rows and 15 columns(dimension)

# Step 5:
Data Formatting and Data Normalization: Summarize the types of variables by checking the data types (i.e., character, numeric, integer, factor, and logical) of the variables in the data set. If variables are not in the correct data type, apply proper type conversions.

Integer Variables:

* survived: Represents whether a passenger survived (0 or 1).
* pclass: Represents the passenger class (1, 2, or 3).
* sibsp: Represents the number of siblings/spouses aboard the Titanic.
* parch: Represents the number of parents/children aboard the Titanic.

Float Variables:

* age: Represents the age of the passenger.
* fare: Represents the fare paid by the passenger.

Object Variables:

* sex: Represents the gender of the passenger (male or female).
* embarked: Represents the port of embarkation (C, Q, or S).
* who: Represents the classification of the passenger (man, woman, or child).
* embark_town: Represents the town of embarkation.
* alive: Represents whether the passenger survived (yes or no).

Boolean Variables:

* adult_male: Represents whether the passenger is an adult male.
* alone: Represents whether the passenger is traveling alone.

Categorical Variables:

* class: Represents the passenger class as a categorical variable.
* deck: Represents the deck on which the passenger's cabin was located as a categorical variable.

From the above summary of the data types of variables we can apply type conversions for:
Convert sex, embarked, who, embark_town, and alive from object to categorical.

In [ ]:
# Convert object variables to categorical
df[['sex', 'who', 'class']] = df[['sex', 'who', 'class']].astype('category')


<ipython-input-53-c4d6d28fc0d3>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df[['sex', 'who', 'class']] = df[['sex', 'who', 'class']].astype('category')


#Step 6: Turn categorical variables into quantitative variables in Python.

Converting categorical variables into quantitative variables involves transforming categories into numerical representations that can be used in mathematical models or analyses. This conversion is necessary because many machine learning algorithms and statistical techniques require numerical input.

In [ ]:
# Perform one-hot encoding on categorical variables
encoded = pd.get_dummies(df, columns=['sex', 'who', 'class'])

# Display the first few rows of the encoded DataFrame
print(encoded.head())

   survived  pclass   age  sibsp  parch     fare embarked  adult_male deck  \
0         0       3  22.0      1      0   7.2500        S        True  NaN   
1         1       1  38.0      1      0  71.2833        C       False    C   
2         1       3  26.0      0      0   7.9250        S       False  NaN   
3         1       1  35.0      1      0  53.1000        S       False    C   
4         0       3  35.0      0      0   8.0500        S        True  NaN   

   embark_town alive  alone  sex_female  sex_male  who_child  who_man  \
0  Southampton    no  False       False      True      False     True   
1    Cherbourg   yes  False        True     False      False    False   
2  Southampton   yes   True        True     False      False    False   
3  Southampton   yes  False        True     False      False    False   
4  Southampton    no   True       False      True      False     True   

   who_woman  class_First  class_Second  class_Third  
0      False        False         Fal

Explanation:

pd.get_dummies(): This function is used to perform one-hot encoding on categorical variables. It creates binary columns for each category in the specified columns.

columns=['sex', 'embarked', 'who', 'embark_town', 'alive', 'class', 'deck']: This parameter specifies the columns to encode. Replace these column names with the categorical variables in your DataFrame.

df_encoded: This DataFrame contains the original data with categorical variables replaced by one-hot encoded columns.

By applying one-hot encoding, you transform categorical variables into a format that can be used as input in machine learning models or other numerical analyses. Each category is represented by a binary column, making it suitable for quantitative analysis.

In [ ]:
from sklearn.preprocessing import LabelEncoder

# Initialize LabelEncoder
label_encoder = LabelEncoder()
# Apply label encoding to each categorical variable using .loc
for column in ['sex', 'who', 'class']:
    df.loc[:, column] = label_encoder.fit_transform(df[column])

# Display the first few rows of the DataFrame with encoded categorical variables
print(df.head(10))

   survived  pclass  sex   age  sibsp  parch     fare embarked  class  who  \
0         0       3    1  22.0      1      0   7.2500        S      2    1   
1         1       1    0  38.0      1      0  71.2833        C      0    2   
2         1       3    0  26.0      0      0   7.9250        S      2    2   
3         1       1    0  35.0      1      0  53.1000        S      0    2   
4         0       3    1  35.0      0      0   8.0500        S      2    1   
5         0       3    1   NaN      0      0   8.4583        Q      2    1   
6         0       1    1  54.0      0      0  51.8625        S      0    1   
7         0       3    1   2.0      3      1  21.0750        S      2    0   
8         1       3    0  27.0      0      2  11.1333        S      2    2   
9         1       2    0  14.0      1      0  30.0708        C      1    0   

   adult_male deck  embark_town alive  alone  
0        True  NaN  Southampton    no  False  
1       False    C    Cherbourg   yes  False  


Three methods to convert categorical into numerical


1.   One-hot encoding
2.   Label encoding
3.   Using astype()
astype() is a method available for Series and DataFrame objects in pandas.
df['sex'] = df['sex'].**astype('category')**.cat.codes

cat.codes: This attribute returns the category codes for each category in the categorical variable. The categories are automatically assigned integer codes starting from 0.

# Convert encoded columns back to categorical data type
df['sex'] = df['sex'].astype('category')